File: python/tests/test_backbone.py
from __future__ import annotations

import importlib
import sys
import unittest
from pathlib import Path

SRC = Path(__file__).resolve().parents[1] / "src"
sys.path.insert(0, str(SRC))

from pinn_m3dc1 import MODULE_SPECS, ProjectState, nominal_startup_order
from pinn_m3dc1.common.registry import validate_registry
from pinn_m3dc1.common.scaffold import ScaffoldModule


class BackboneTests(unittest.TestCase):
    def test_registry_is_valid(self) -> None:
        validate_registry()
        self.assertEqual(len(MODULE_SPECS), 7)

    def test_nominal_order_is_numbered(self) -> None:
        self.assertEqual(nominal_startup_order(), tuple(f"module_{i}" for i in range(1, 8)))

    def test_shared_state_contract(self) -> None:
        state = ProjectState()
        module = ScaffoldModule(module_id="module_test", produced_marker="test_product")
        module.run(state)
        self.assertTrue(state.fields["test_product"])
        self.assertEqual(state.history[-1]["module"], "module_test")

    def test_missing_product_fails_loudly(self) -> None:
        state = ProjectState()
        module = ScaffoldModule(module_id="module_test", required_products=("missing",))
        with self.assertRaises(KeyError):
            module.run(state)

    def test_all_module_scaffolds_are_addressable(self) -> None:
        for spec in MODULE_SPECS:
            solver = importlib.import_module(f"pinn_m3dc1.{spec.package}.solver")
            module = solver.create()
            self.assertEqual(module.module_id, spec.id)


if __name__ == "__main__":
    unittest.main()

File: matlab/tests/test_backbone.m
function tests = test_backbone
%TEST_BACKBONE Contract tests for the MATLAB project scaffold.
tests = functiontests(localfunctions);
end

function setupOnce(testCase)
testsDir = fileparts(mfilename('fullpath'));
matlabRoot = fileparts(testsDir);
sourceDir = fullfile(matlabRoot, 'src');
addpath(sourceDir);
testCase.TestData.sourceDir = sourceDir;
end

function teardownOnce(testCase)
rmpath(testCase.TestData.sourceDir);
end

function testRegistryIsValid(testCase)
pinn_m3dc1.common.validate_registry();
specs = pinn_m3dc1.common.registry();
verifyEqual(testCase, numel(specs), 5);
end

function testNominalOrderIsNumbered(testCase)
actual = pinn_m3dc1.common.nominal_startup_order();
expected = {'module_1', 'module_2', 'module_3', 'module_4', 'module_5'};
verifyEqual(testCase, actual, expected);
end

function testSharedStateContract(testCase)
state = pinn_m3dc1.common.ProjectState();
module = pinn_m3dc1.common.ScaffoldModule( ...
    'module_test', {}, 'test_product');
module.run(state);
verifyTrue(testCase, state.fields.test_product);
verifyEqual(testCase, state.history{end}.module, 'module_test');
end

function testMissingProductFailsLoudly(testCase)
state = pinn_m3dc1.common.ProjectState();
module = pinn_m3dc1.common.ScaffoldModule( ...
    'module_test', {'missing'}, 'unused');
verifyError(testCase, @() module.run(state), ...
    'pinn_m3dc1:MissingProducts');
end

function testAllModuleScaffoldsAreAddressable(testCase)
specs = pinn_m3dc1.common.registry();
for i = 1:numel(specs)
    factory = str2func(sprintf('pinn_m3dc1.%s.solver', specs(i).package));
    module = factory();
    verifyEqual(testCase, module.moduleId, specs(i).id);
end
end
